doc/design/design_sqlflow_parser.md

```diff
@@ -0,0 +1,149 @@
+1. Distinguishing a SQL statement between standard SQL and extended SQL becomes trivial (i.e., if the third party parser can't parse the statement alone, then it is an extended SQL statement).
+1. Splitting a SQL program into multiple SQL statements becomes trivial.
+
+Cons:
+1. The error position `p` raise by TPP might be inaccurate. However, we haven't found any counterexample in our experiments at #1046 and #1103.
+
+Based on the above pros and cons, we choose the TPP solution.
+
+### gRPC vs. Local Command Line
+
+SQLFlow is written in Go, third party parsers for Calcite/HiveQL are written in Java. So we need to call the third parse via either gRPC or local command line.
+
+We prefer the local command line for the following reason.
+1. We can avoid setting up an additional gRPC server, which makes the testing/deployment simpler. And because there is no downstream call for the parser, wrapping the parser as a service might be overkill.
+1. We can guarantee the isolation between parsing requests. The gPRC server may fail due to an unexpected request, meaning one request can affect the other.
+
+## Implementation Details
+
+### Change `TRAIN/PREDICT` to `TO TRAIN/PREDICT`
+
+SQLFlow syntax should be compatible with the standard SQL syntax. During our earlier experiments, we found using `TRAIN` along to denote the SQL extension is not enough. In cases like
+1. `select * from mytable train dnn ...`, the SQL syntax will treat `train` as an alias of `mytable`.
+1. `select * from mytable train` where we eliminate the model name because of the automatic model selection.
+
+These cases lead us to the motivation of changing `train` to `to train`. All [MySQL](https://dev.mysql.com/doc/refman/5.5/en/keywords.html), [MaxCompute](https://www.alibabacloud.com/help/doc-detail/27872.htm?spm=a2c63.p38356.879954.6.631c5095hrjubf#concept-zxk-v5f-vdb), and [Hive](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL) consider `TO` as a reversed keyword. And its usage is subjected in the following scenario.
+1. Hive: `ALTER TABLE table_name RENAME TO new_table_name;`.
+1. MySQL: `TO` is used in [`GRANT`](https://dev.mysql.com/doc/refman/8.0/en/grant.html) and [`RENAME`](https://dev.mysql.com/doc/refman/8.0/en/rename-table.html).
+1. MaxCompute:
+    1. [`ALTER TABLE table_name RENAME TO new_table_name;`](https://www.alibabacloud.com/help/doc-detail/73768.html?spm=a2c5t.11065259.1996646101.searchclickresult.5afd4bd7qECSMQ)
+    1. [`alter table table_name changeowner to 'ALIYUN$xxx@aliyun.com';`](https://www.alibabacloud.com/help/doc-detail/73768.html?spm=a2c5t.11065259.1996646101.searchclickresult.5afd4bd7qECSMQ)
+
+So Given a SQL statement `SELECT ... TO TRAIN`, the third-party parser should raise error precisely at the position of `TO`.
+
+### Third Party Parser API
+
+The third-party parser provides a function `ParseAndSplit`, it takes a SQL program and returns three elements: `statements`, `position`, and `error`.
```

Comment:
Do the TPP API also need an API to return the AST or just columns to verify the SELECT columns and COLUMN statement 
 in extended SQL?